### PR TITLE
fix(module-tools): tsconfig parse & resolution should support `extends` field

### PR DESCRIPTION
**Pull request by kirainmoe**
## Summary

This PR tried to support resolve extended tsconfig when parsing project tsconfig.json for module-tools.

## Checklist

- [x] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [x] I have added tests to cover my changes.

### CODE REVIEW

**Comment by 10Derozan** [comment]
Nice, please remove deepmerge, and import lodash.merge from @modern-js/utils.

**Comment by kirainmoe** [comment]
> Nice, please remove deepmerge, and import lodash.merge from @modern-js/utils.

I tried this, but it seems that lodash merge can not handle the situation when `target` has a field that `source` does not have. For example:

```
// tsconfig.base.json
{
    "compilerOptions": {
        "declaration": true
    }
}

// tsconfig.json
{
    "extends": "tsconfig.base.json",
    "compilerOptions": {
        "emitDecoratorMetadata": true
    }
}
```

If I merge `tsconfig.json` into `tsconfig.base.json` merge result of "compilerOptions" will only have `declaration` because `emitDecoratorMetadata` does not exist on `tsconfig.base.json`'s `compilerOptions`.

I found some references about this: https://github.com/lodash/lodash/issues/5242

So I used `deepmerge` like other projects in this monorepo did.

Is there any suggestion? Should I implement a deep merge by myself?

**Comment by kirainmoe** [comment]
I saw ut-mac failed to pass so I rebased main branch :(

**Comment by 10Derozan** [comment]
> > Nice, please remove deepmerge, and import lodash.merge from @modern-js/utils.
> 
> I tried this, but it seems that lodash merge can not handle the situation when `target` has a field that `source` does not have. For example:
> 
> ```
> // tsconfig.base.json
> {
>     "compilerOptions": {
>         "declaration": true
>     }
> }
> 
> // tsconfig.json
> {
>     "extends": "tsconfig.base.json",
>     "compilerOptions": {
>         "emitDecoratorMetadata": true
>     }
> }
> ```
> 
> If I merge `tsconfig.json` into `tsconfig.base.json` merge result of "compilerOptions" will only have `declaration` because `emitDecoratorMetadata` does not exist on `tsconfig.base.json`'s `compilerOptions`.
> 
> I found some references about this: [lodash/lodash#5242](https://github.com/lodash/lodash/issues/5242)
> 
> So I used `deepmerge` like other projects in this monorepo did.
> 
> Is there any suggestion? If import `deepmerge` is not allowed, should I implement a deep merge by myself?

It can not be solved by mergeWith? I have not checked it.
If lodash.merge can not solve this, use deepmerge.

**Comment by kirainmoe** [comment]
Sorry, I missed the comments in the issue, `mergeWith` did solve this. I replaced deepmerge with `mergeWith`: https://github.com/web-infra-dev/modern.js/pull/5427/commits/376c0c1c534dc949373a96168e6bf7574a25b8b9 and https://github.com/web-infra-dev/modern.js/pull/5427/commits/a38a93877db5d5a720045d7fd095225cad5a61aa .

Also I did more that and fix an edge case: when "extends" field extending from a npm package root, the previous implementation failed to resolve it, so I made some modification to the resolution logic, please review it in this commit :) https://github.com/web-infra-dev/modern.js/pull/5427/commits/2e2221c8dd04c9c91757c5220d0c8b1d13a7c46b

File: .changeset/giant-bananas-smile.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@modern-js/module-tools': patch
+---
+
+支持解析 tsconfig extends
```
[the remark this answers is not in the document]
Comment (kirainmoe) [reply .changeset/giant-bananas-smile.md]: done

**Review by 10Derozan — APPROVED** [review]
Good job

**Pull request merged**